readme.md
# Spam-Email-Detection-Toolbox

In the following, we will design a SPAM email detection toolbox. First we will compose the
emails, then we will detect spam. Do not hardcode values/logic when not specified.

# Part A

1. Create a text file named student_name_list.txt of the names of 10 PC503 students manually,
keeping one name in each line. Note: this is the only file that is to be created manually. Files
in the subsequent questions must not be created manually.

2. Write a Python code main_data.py to do the following tasks. Each task must be written in a
function with a suitable function name. Depending on the task, the function may or may not
return anything.

3. Read the file About-DAIICT.txt and create four new files named History.txt,
Environment.txt, Recognition.txt and Accreditation.txt by inserting the content of each section
as given in About-DAIICT.txt.

4. Create a list named student_name_list of the 10 names by reading the text file
student_name_list.txt.

5. Randomly generate 15 files with file names: email-1.txt to email-20.txt. So, the name of the
13 th file need not be email-13.txt, it can also be email-19.txt or email-5.txt. Content of each
file (10 lines) should be as follows:
a) First line should have time, date, month, year, email Id in the following format:
Received at 17:23 hrs on October 7, 2018 from fname_lname@daiict.ac.in
All entries must be randomly inserted and email Ids can be created from random elements of
student_name_list. Obviously, there will be at least one or more files containing the same
email Id since the number of students is 10 (<15).
b) Second line should be blank.
c) Third line to the tenth line must be the lines from the four above mentioned text files
in random order such that (i) there is at least one line from each file and (ii) the lines
taken from History.txt can include only one of the four sentences containing “first
wave”, “second wave”, “third wave”, “fourth wave”. There is no restriction on other
sentences in this text file.

6. Randomly generate 4 more files using the above mentioned logic with two modifications:
a) File name should be the unused numbers between 1 to 20.
b) Email Ids must be taken from the following list: email_ids =[ name1@gmail.com,
A_X_y@yahoo.co.in, nm123@rediff.com, nam_4_e@160.com].

7. Randomly generate 1 more file using the above mentioned logic with three modifications:
a) File name should be the unused number between 1 to 20.
b) Email Id is pc_503@daiict.ac.in
c) The lines taken from History.txt cannot include any one of the four sentences
containing “first wave”, “second wave”, “third wave”, “fourth wave”. There is no
restriction on other sentences in this text file.

# Part B

8. Write a Python code main_test.py to do the following tasks. Each task must be written in a
function with a suitable function name. Depending on the task, the function may or may not
return anything.

9. Print the email Ids whose files do not have any “wave” information. In the current exercise,
there is only one such entry.

10. The users not having daiict email Id or without any “wave” information must be spammed.
Rename the corresponding files as spam1.txt, spam2.txt, ... and update their contents by
adding a line at the beginning: “This email has been categorized as spam”.

11. Create ordered_names.txt containing the user names (non-spam type) in the descending
order of date-time etc., i.e. most recent first. One name in each line. Keep duplicate names.

12. Create ordered_names_wave1.txt containing the user names (non-spam type) belonging to
the “first wave” group in the descending order of date-time etc., i.e. most recent first. No
duplicate names and one name in each line. Similarly create files for the “second wave”,
“third wave”, “fourth wave” groups.

13. Plot the number of users in each group and number of spam emails using a bar chart.
Note: Randomize the codes so that every time they are executed, different output is expected.



History.txt
History

Dhirubhai Ambani Institute of Information and Communication Technology (DA-IICT), Gandhinagar represents Wave-4 of educational innovation in Gujarat.

The first wave was the nationalist wave and led to Gandhian experiments in education including Nai Talim. 
The Gujarat Vidyapith established in 1920 was a hybrid model of a University based on Gandhian principles.

The second wave led to the establishment of a whole range of educational institutions in Gujarat, mainly private colleges.

The third wave, an inspired one, was spearheaded by the Industrialist Kasturbhai Lalbhai and the Scientist Vikram Sarabhai and led to the establishment of a network of national institutes of international renown. 
A whole array of remarkable intellectuals provided the leadership of these Institutes.

It was in the fourth wave when the focus shifted to higher education and private participation. One of the Institutes created during this period is DA-IICT in 2001. 
It is the only advanced research and teaching Institute named after the Late Dhirubhai Ambani, the founder of Reliance Group.


main_test.py
import os
from datetime import datetime
from collections import OrderedDict

def emails_without_wave_information(email_files):
    spam_emails = {}
    for email in email_files:
        with open(email, 'r') as s:
            lines = s.readlines()
            if lines[2].find("wave") == -1:
                spam_emails[email]=lines[0].split()[-1]
    return spam_emails

def emails_without_daiictid(email_files):
    spam_emails = {}
    for email in email_files:
        with open(email, 'r') as s:
            lines = s.readlines()
            if lines[0].split()[-1].find("daiict.ac.in") == -1:
                spam_emails[email]=lines[0].split()[-1]
    return spam_emails

def rename_files(spam_emails):
    i=1
    for email in spam_emails:
        with open(email,'r+') as s:
            content = s.read()
            s.seek(0)
            s.write("This email has been categorized as spam\n"+content)
        os.rename(email,"spam"+str(i)+".txt")    
        i += 1

def time_to_timestamp(t_d):
    months = {
        'January':1,
        'February':2, 
        'March':3,
        'April':4, 
        'May':5, 
        'June':6, 
        'July':7, 
        'August':8, 
        'September':9, 
        'October': 10, 
        'November':11, 
        'December':12
    }

    hr, minute = map(int, t_d[0].split(':'))  
    month = months[t_d[-3]] 
    day =  int(t_d[-2][:-1]) 
    year = int(t_d[-1])

    return datetime(year, month, day, hr, minute, 0, 0) 

def gen_timestamp_name_dict(to_be_written, line):
    content = line.split()
    t_d, name = time_to_timestamp(content[2:8]), content[-1]
    if to_be_written.get(t_d):
        to_be_written[t_d].append(name)
    else:
        to_be_written[t_d] = [name.split('@')[0].replace("_", " ")]

# trial function - to remove redundant names mapped on different timestamps - can be optimized
def get_unique_dict(to_be_written):
    reverse_dict = dict()
    to_be_returned = dict()
    for k, v in to_be_written.items():
        for name in v:
            if reverse_dict.get(name):
                reverse_dict[name].append(k)
            else:
                reverse_dict[name] = [k]

    for k, v in reverse_dict.items():
        reverse_dict[k]= sorted(v)[-1]
    
    for k,v in reverse_dict.items():
        if to_be_returned.get(v):
            to_be_returned[v].append(k)
        else:
            to_be_returned[v] = [k]

    return to_be_returned



def write_ordered_dict(file, to_be_written, asc):
    if asc:
        Ordered = OrderedDict(sorted(to_be_written.items()))
    else:
        Ordered = OrderedDict(reversed(sorted(to_be_written.items())))
    
    with open(file, 'w') as ord:
        for k,v in Ordered.items():
            for name in v:
                ord.write(str(k)  + " " + name + "\n")    


def generate_ordered_names(lst):
    to_be_written = dict()
    for file in lst:
        with open(file, 'r') as fl:
            line = fl.readline()
            gen_timestamp_name_dict(to_be_written, line)

    write_ordered_dict('Ordered_names.txt', to_be_written, True)

            
def generate_ordered_names_wave(lst):
    to_be_written_f1 = dict()
    to_be_written_f2 = dict()
    to_be_written_f3 = dict()
    to_be_written_f4 = dict()
    for file in lst:
        with open(file, 'r') as f1:
            line = f1.readline()
            content = f1.read()
            if content.find("first wave") != -1:
                gen_timestamp_name_dict(to_be_written_f1, line)
            elif content.find("second wave") != -1:
                gen_timestamp_name_dict(to_be_written_f2, line)
            elif content.find("third wave") != -1:
                gen_timestamp_name_dict(to_be_written_f3, line)
            elif content.find("fourth wave") != -1:
                gen_timestamp_name_dict(to_be_written_f4, line)


    write_ordered_dict("ordered_names_wave1.txt", get_unique_dict(to_be_written_f1), False)
    write_ordered_dict("ordered_names_wave2.txt", get_unique_dict(to_be_written_f2), False)
    write_ordered_dict("ordered_names_wave3.txt", get_unique_dict(to_be_written_f3), False)
    write_ordered_dict("ordered_names_wave4.txt", get_unique_dict(to_be_written_f4), False)


email_files = [f'email-{i}.txt' for i in range(1,21)]
spam_emails = emails_without_wave_information(email_files)
print ("Emails that not contain wave information are :")
for email in spam_emails:
  print (spam_emails[email])
spam_emails.update(emails_without_daiictid(email_files))
print('Spam:', spam_emails.keys())
rename_files(spam_emails)

non_spam_emails = list(set(email_files).difference(set(spam_emails.keys())))
print('Non spam:', non_spam_emails)

generate_ordered_names(non_spam_emails)
generate_ordered_names_wave(non_spam_emails)

main_data.py
import random
import os
from randomtimestamp import randomtimestamp

random.seed(53)

def random_time():
    timestamp=randomtimestamp(2010,text=False)
    month = timestamp.strftime("%B") 
    year = timestamp.strftime("%Y")
    date = int(timestamp.strftime("%d"))     
    time = timestamp.strftime("%H")+":"+timestamp.strftime("%M")
    return str("Recieved at "+time+" hrs on "+month+" "+str(date)+", "+year+" from ")

def write_file(file_name, content):
    with open(file_name, 'w') as op:
        op.writelines(content) 

def filter_data(constraint, lst):
    string_filter = lambda string : False if string == "\n" else True
    lst = list(filter(constraint, list(filter(string_filter, lst))))
    return lst

def details_about_DAIICT():
    with open('./About-DAIICT.txt', 'r') as ad:
        lines       = ad.readlines()

        structure = {'start':0,'end': 0}

        history     = structure.copy()
        environment = structure.copy()
        recognition = structure.copy()
        accreditation = structure.copy()

        i = 0
        while i < len(lines):
            if lines[i] == 'History\n':
                history['start'] = i

            elif lines[i] == 'Environment\n':
                history['end'] = i-1
                environment['start'] = i

            elif lines[i] == 'Recognition\n':
                environment['end'] = i-1
                recognition['start'] = i

            elif lines[i] == 'Accreditation\n':
                recognition['end'] = i-1
                accreditation['start'] = i

            i += 1

        history_r = lines[history['start']:history['end']+1]
        write_file('History.txt', history_r)

        environment_r = lines[environment['start']:environment['end']+1]
        write_file('Environment.txt', environment_r)
        
        recognition_r = lines[recognition['start']:recognition['end']+1]
        write_file('Recognition.txt', recognition_r)

        accreditation_r = lines[accreditation['start']:]
        write_file('Accreditation.txt', accreditation_r)

        return {
            'history'      :     history_r, 
            'environment'  : environment_r, 
            'recognition'  : recognition_r, 
            'accreditation':accreditation_r
        }


def emails(student_name_list, random_email_files, content):

    history = content['history']
    environment = content['environment']
    recognition = content['recognition']
    accreditation = content['accreditation']

    for random_email_file in random_email_files:
        with open(random_email_file, 'w') as remail:

            history_s = random.choice(history)
            env_s = random.choice(environment)
            rec_s = random.choice(recognition)
            acc_s = random.choice(accreditation)

            remail.writelines([
                    f'{random_time()}{random.choice(student_name_list)}\n',
                    '\n',
                    history_s,
                    env_s,
                    rec_s,
                    acc_s
                ]
            )


def emails_original(content):

    content_l = content.copy()

    student_name_list = []
    with open('student_names_list.txt', 'r') as s:
        student_name_list = s.readlines()

    student_name_list = list(
        map(
            lambda string: string.replace("\n", "").replace(" ", "_") + "@daiict.ac.in", 
            student_name_list
        ))

    random_email_files = [
        f'email-{i}.txt' for i in random.sample(range(1, 21), 15)
    ]
    
    ready_to_go = lambda string: True
    content_l['environment']   = filter_data(ready_to_go, content_l['environment'])
    content_l['recognition']   = filter_data(ready_to_go, content_l['recognition'])
    content_l['accreditation'] = filter_data(ready_to_go, content_l['accreditation'])

    history_constraint = lambda string : True if (
            string.startswith('The first wave') or 
            string.startswith('The second wave') or 
            string.startswith('The third wave') or
            string.startswith('It was in the fourth wave')) else False

    content_l['history'] = filter_data(history_constraint, content_l['history'])
    emails(student_name_list, random_email_files, content_l)

    return random_email_files


def emails_with_modifications(content, files):

    content_l = content.copy()
 
    files = set(i for i in range(1,21)).difference(set(int(email.split('-')[1].split('.')[0]) for email in files))

    email_file_list = random.sample([f'email-{file}.txt' for file in files], 4)
    
    student_name_list = ['name1@gmail.com', 'A_X_y@yahoo.co.in', 'nm123@rediff.com', 'nam_4_e@160.com']

    ready_to_go = lambda string: True
    content_l['environment']   = filter_data(ready_to_go, content_l['environment'])
    content_l['recognition']   = filter_data(ready_to_go, content_l['recognition'])
    content_l['accreditation'] = filter_data(ready_to_go, content_l['accreditation'])

    history_constraint = lambda string : True if (
            string.startswith('The first wave') or 
            string.startswith('The second wave') or 
            string.startswith('The third wave') or
            string.startswith('It was in the fourth wave')) else False

    content_l['history'] = filter_data(history_constraint, content_l['history'])

    emails(student_name_list, email_file_list, content_l)

    return email_file_list

def email_with_another_modifications(content, files):

    content_l = content.copy()
 
    files = set(i for i in range(1,21)).difference(set(int(email.split('-')[1].split('.')[0]) for email in files))

    email_file_list = random.sample([f'email-{file}.txt' for file in files], 1)
    
    student_name_list = ['pc_503@daiict.ac.in']

    ready_to_go = lambda string: True
    content_l['environment']   = filter_data(ready_to_go, content_l['environment'])
    content_l['recognition']   = filter_data(ready_to_go, content_l['recognition'])
    content_l['accreditation'] = filter_data(ready_to_go, content_l['accreditation'])

    history_constraint = lambda string : True if not (
            string.startswith('The first wave') or 
            string.startswith('The second wave') or 
            string.startswith('The third wave') or
            string.startswith('It was in the fourth wave')) else False

    content_l['history'] = filter_data(history_constraint, content_l['history'])

    emails(student_name_list, email_file_list, content_l)
    return email_file_list

content = details_about_DAIICT()
random_email_files = emails_original(content)
random_email_files = emails_with_modifications(content, random_email_files) + random_email_files
random_email_files = email_with_another_modifications(content, random_email_files) + random_email_files



